File: backend/app/api/voice_search.py
from fastapi import APIRouter, UploadFile, File, HTTPException, Depends
from sqlalchemy.orm import Session
from pydantic import BaseModel

from app.database.connection import get_db
from app.middleware.auth import get_current_user
from app.services.voice_search_service import VoiceSearchService

router = APIRouter()

voice_service = VoiceSearchService()


# ---------- Schemas ----------

class VoiceTextRequest(BaseModel):
    text: str


# ---------- Speech to Text ----------

@router.post("/speech-to-text")
async def speech_to_text(
        audio: UploadFile = File(...)
):

    try:
        return voice_service.speech_to_text(
            audio
        )

    except Exception as e:

        raise HTTPException(
            status_code=500,
            detail=str(e)
        )


# ---------- Voice Product Search ----------

@router.post("/search")
async def voice_search(
        audio: UploadFile = File(...),
        db: Session = Depends(get_db)
):

    try:

        return voice_service.voice_search(
            db=db,
            audio=audio
        )

    except Exception as e:

        raise HTTPException(
            status_code=500,
            detail=str(e)
        )


# ---------- Search By Text ----------

@router.post("/text-search")
def text_search(
        request: VoiceTextRequest,
        db: Session = Depends(get_db)
):

    return voice_service.text_search(
        db=db,
        text=request.text
    )


# ---------- Voice Recommendations ----------

@router.post("/recommend")
async def voice_recommendation(
        audio: UploadFile = File(...),
        db: Session = Depends(get_db),
        current_user=Depends(get_current_user)
):

    return voice_service.voice_recommendation(
        db=db,
        user_id=current_user.id,
        audio=audio
    )


# ---------- Category Search ----------

@router.post("/category")
async def category_search(
        audio: UploadFile = File(...)
):

    return voice_service.category_search(
        audio
    )


# ---------- Voice Commands ----------

@router.post("/command")
def voice_command(
        request: VoiceTextRequest
):

    return voice_service.voice_command(
        request.text
    )


# ---------- Search History ----------

@router.get("/history")
def voice_history(
        current_user=Depends(get_current_user)
):

    return voice_service.voice_history(
        current_user.id
    )


# ---------- Trending Queries ----------

@router.get("/trending")
def trending_voice_queries():

    return voice_service.trending_voice_queries()


# ---------- AI Assistant ----------

@router.post("/assistant")
def voice_assistant(
        request: VoiceTextRequest
):

    return voice_service.voice_assistant(
        request.text
    )


# ---------- Health Check ----------

@router.get("/health")
def health_check():

    return {
        "status": "healthy",
        "service": "voice_search"
    }

File: frontend/src/services/voiceSearch.ts
import { api } from './api';
import type { Product } from './product';

export interface VoiceCommandResponse {
  command: string;
  action: string;
  parameters: { [key: string]: any };
  message: string;
}

export const voiceSearchService = {
  async speechToText(audioBlob: Blob): Promise<{ text: string }> {
    const formData = new FormData();
    formData.append('audio', audioBlob, 'voice.wav');
    const res = await api.post<{ text: string }>('/api/voice-search/speech-to-text', formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return res.data;
  },

  async voiceSearch(audioBlob: Blob): Promise<Product[]> {
    const formData = new FormData();
    formData.append('audio', audioBlob, 'voice.wav');
    const res = await api.post<Product[]>('/api/voice-search/search', formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return res.data;
  },

  async textSearch(text: string): Promise<Product[]> {
    const res = await api.post<Product[]>('/api/voice-search/text-search', { text });
    return res.data;
  },

  async voiceRecommend(audioBlob: Blob): Promise<Product[]> {
    const formData = new FormData();
    formData.append('audio', audioBlob, 'voice.wav');
    const res = await api.post<Product[]>('/api/voice-search/recommend', formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return res.data;
  },

  async categorySearch(audioBlob: Blob): Promise<{ category: string; products: Product[] }> {
    const formData = new FormData();
    formData.append('audio', audioBlob, 'voice.wav');
    const res = await api.post<{ category: string; products: Product[] }>('/api/voice-search/category', formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return res.data;
  },

  async runVoiceCommand(text: string): Promise<VoiceCommandResponse> {
    const res = await api.post<VoiceCommandResponse>('/api/voice-search/command', { text });
    return res.data;
  },

  async getVoiceHistory(): Promise<Array<{ id: number; query: string; created_at: string }>> {
    const res = await api.get<Array<{ id: number; query: string; created_at: string }>>('/api/voice-search/history');
    return res.data;
  },

  async getTrendingQueries(): Promise<string[]> {
    const res = await api.get<string[]>('/api/voice-search/trending');
    return res.data;
  },

  async askVoiceAssistant(text: string): Promise<{ reply: string; data?: any }> {
    const res = await api.post<{ reply: string; data?: any }>('/api/voice-search/assistant', { text });
    return res.data;
  }
};
